File: sdk/kotlin/src/main/kotlin/com/loomup/client/TableQuery.kt
package com.loomup.client

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/** Fluent table accessor: CRUD + realtime for one table name. */
class TableQuery internal constructor(
    private val client: LoomupClient,
    private val table: String,
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        isLenient = true
    }

    suspend fun select(
        where: Map<String, WhereValue>? = null,
        filter: Map<String, Map<String, WhereValue>>? = null,
        select: List<String>? = null,
        sort: String? = null,
        limit: Int? = null,
        offset: Int? = null,
        cursor: String? = null,
    ): ListResult {
        val items = mutableListOf<Pair<String, String>>()
        if (cursor != null) items.add("cursor" to cursor)
        if (limit != null) items.add("limit" to limit.toString())
        if (offset != null) items.add("offset" to offset.toString())
        if (sort != null) items.add("sort" to sort)
        if (!select.isNullOrEmpty()) items.add("select" to select.joinToString(","))
        if (where != null) {
            for ((k, v) in where) {
                items.add("where[$k]" to v.queryString)
            }
        }
        if (filter != null) {
            for ((field, operations) in filter) {
                for ((operation, value) in operations) {
                    val wireOperation = when (operation) {
                        "isNull" -> "is_null"
                        "startsWith" -> "starts_with"
                        else -> operation
                    }
                    items.add("filter[$field][$wireOperation]" to value.queryString)
                }
            }
        }

        var path = "/api/${encodeURIComponent(table)}"
        if (items.isNotEmpty()) {
            val query = items.joinToString("&") { (k, v) ->
                "${encodeQueryComponent(k)}=${encodeQueryComponent(v)}"
            }
            path += "?$query"
        }

        val env: ListEnvelope = client.requestJson("GET", path)
        return ListResult(data = env.data, meta = env.meta)
    }

    suspend fun get(id: String): Map<String, JsonValue> {
        val path = "/api/${encodeURIComponent(table)}/${encodeURIComponent(id)}"
        val env: DataEnvelope<Map<String, JsonValue>> = client.requestJson("GET", path)
        return env.data
    }

    suspend fun get(id: Int): Map<String, JsonValue> = get(id.toString())

    suspend fun insert(row: Map<String, JsonValue>): Map<String, JsonValue> {
        val body = json.encodeToString(row).toByteArray(StandardCharsets.UTF_8)
        val path = "/api/${encodeURIComponent(table)}"
        val env: DataEnvelope<Map<String, JsonValue>> = client.requestJson("POST", path, body)
        return env.data
    }

    suspend fun update(id: String, patch: Map<String, JsonValue>): Map<String, JsonValue> {
        val body = json.encodeToString(patch).toByteArray(StandardCharsets.UTF_8)
        val path = "/api/${encodeURIComponent(table)}/${encodeURIComponent(id)}"
        val env: DataEnvelope<Map<String, JsonValue>> = client.requestJson("PATCH", path, body)
        return env.data
    }

    suspend fun update(id: Int, patch: Map<String, JsonValue>): Map<String, JsonValue> =
        update(id.toString(), patch)

    suspend fun delete(id: String): Map<String, JsonValue> {
        val path = "/api/${encodeURIComponent(table)}/${encodeURIComponent(id)}"
        val env: DataEnvelope<Map<String, JsonValue>> = client.requestJson("DELETE", path)
        return env.data
    }

    suspend fun delete(id: Int): Map<String, JsonValue> = delete(id.toString())

    fun subscribe(
        rowId: String? = null,
        handler: SubscribeHandler,
    ): Unsubscribe {
        return client.subscribeTable(table = table, rowId = rowId, handler = handler)
    }

    /** Awaitable subscribe — resolves when the server acknowledges (`subscribed` frame). */
    suspend fun subscribeReady(
        rowId: String? = null,
        timeoutMs: Int = 5000,
        handler: SubscribeHandler,
    ): Unsubscribe {
        return client.subscribeTableReady(
            table = table,
            rowId = rowId,
            timeoutMs = timeoutMs,
            handler = handler,
        )
    }

    private fun encodeQueryComponent(value: String): String {
        return URLEncoder.encode(value, StandardCharsets.UTF_8.name())
            .replace("+", "%20")
    }
}

File: sdk/swift/Sources/Loomup/TableQuery.swift
import Foundation

/// Fluent table accessor: CRUD + realtime for one table name.
public struct TableQuery: Sendable {
    private let client: LoomupClient
    private let table: String

    init(client: LoomupClient, table: String) {
        self.client = client
        self.table = table
    }

    public func select(
        where whereClause: [String: WhereValue]? = nil,
        filter: [String: [String: WhereValue]]? = nil,
        select: [String]? = nil,
        sort: String? = nil,
        limit: Int? = nil,
        offset: Int? = nil,
        cursor: String? = nil
    ) async throws -> ListResult {
        var items: [URLQueryItem] = []
        if let cursor { items.append(URLQueryItem(name: "cursor", value: cursor)) }
        if let limit { items.append(URLQueryItem(name: "limit", value: String(limit))) }
        if let offset { items.append(URLQueryItem(name: "offset", value: String(offset))) }
        if let sort { items.append(URLQueryItem(name: "sort", value: sort)) }
        if let select, !select.isEmpty {
            items.append(URLQueryItem(name: "select", value: select.joined(separator: ",")))
        }
        if let whereClause {
            for (k, v) in whereClause {
                items.append(URLQueryItem(name: "where[\(k)]", value: v.queryString))
            }
        }
        if let filter {
            for (field, operations) in filter {
                for (operation, value) in operations {
                    let wireOperation: String
                    switch operation {
                    case "isNull": wireOperation = "is_null"
                    case "startsWith": wireOperation = "starts_with"
                    default: wireOperation = operation
                    }
                    items.append(URLQueryItem(
                        name: "filter[\(field)][\(wireOperation)]",
                        value: value.queryString
                    ))
                }
            }
        }

        var path = "/api/\(encodeURIComponent(table))"
        if !items.isEmpty {
            var components = URLComponents()
            components.queryItems = items
            if let q = components.percentEncodedQuery {
                path += "?\(q)"
            }
        }

        let env: ListEnvelope = try await client.requestJSON(method: "GET", path: path)
        return ListResult(data: env.data, meta: env.meta)
    }

    public func get(_ id: String) async throws -> [String: JSONValue] {
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "GET",
            path: path
        )
        return env.data
    }

    public func get(_ id: Int) async throws -> [String: JSONValue] {
        try await get(String(id))
    }

    public func insert(_ row: [String: JSONValue]) async throws -> [String: JSONValue] {
        let body = try JSONEncoder().encode(row)
        let path = "/api/\(encodeURIComponent(table))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "POST",
            path: path,
            body: body
        )
        return env.data
    }

    public func update(_ id: String, patch: [String: JSONValue]) async throws -> [String: JSONValue] {
        let body = try JSONEncoder().encode(patch)
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "PATCH",
            path: path,
            body: body
        )
        return env.data
    }

    public func update(_ id: Int, patch: [String: JSONValue]) async throws -> [String: JSONValue] {
        try await update(String(id), patch: patch)
    }

    public func delete(_ id: String) async throws -> [String: JSONValue] {
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "DELETE",
            path: path
        )
        return env.data
    }

    public func delete(_ id: Int) async throws -> [String: JSONValue] {
        try await delete(String(id))
    }

    @discardableResult
    public func subscribe(
        rowId: String? = nil,
        handler: @escaping SubscribeHandler
    ) -> Unsubscribe {
        client.subscribeTable(table: table, rowId: rowId, handler: handler)
    }

    /// Awaitable subscribe — resolves when the server acknowledges (`subscribed` frame).
    public func subscribeReady(
        rowId: String? = nil,
        timeoutMs: Int = 5000,
        handler: @escaping SubscribeHandler
    ) async throws -> Unsubscribe {
        try await client.subscribeTableReady(
            table: table,
            rowId: rowId,
            timeoutMs: timeoutMs,
            handler: handler
        )
    }
}
